generator/src/main/java/com/scottlogic/deg/generator/generation/DecisionTreeDataGenerator.java

```diff
@@ -46,30 +51,29 @@ public DecisionTreeDataGenerator(
         GenerationConfig generationConfig) {
 
         monitor.generationStarting(generationConfig);
-        CombinationStrategy partitionCombiner = generationConfig.getCombinationStrategy();
 
         Stream<Stream<DataBag>> partitionedDataBags = treePartitioner
             .splitTreeIntoPartitions(decisionTree)
             .map(treeOptimiser::optimiseTree)
             .map(tree -> generateForPartition(profile, tree, generationConfig));
 
         return partitionCombiner.permute(partitionedDataBags)
-            .map(dataBag -> convertToGeneratedObject(profile, dataBag))
+            .map(dataBag -> convertToGeneratedObject(dataBag, profile))
             .limit(generationConfig.getMaxRows().orElse(GenerationConfig.Constants.DEFAULT_MAX_ROWS))
             .peek(monitor::rowEmitted);
     }
 
     private Stream<DataBag> generateForPartition(Profile profile, DecisionTree tree, GenerationConfig config) {
         FixFieldStrategy fixFieldStrategy = walkerStrategyFactory.getWalkerStrategy(profile, tree, config);
 
-        Stream<DataBagSource> dataBagSources = treeWalker.walk(tree, fixFieldStrategy)
-            .map(dataBagSourceFactory::createDataBagSource);
+        Stream<RowSpec> rowSpecsForPartition = treeWalker.walk(tree, fixFieldStrategy);
 
-        return new ConcatenatingDataBagSource(dataBagSources)
-            .generate(config);
+        return FlatMappingSpliterator.flatMap(
+            rowSpecsForPartition,
+            dataBagSourceFactory::createDataBags);
     }
 
-    private GeneratedObject convertToGeneratedObject(Profile profile, DataBag dataBag) {
+    private GeneratedObject convertToGeneratedObject(DataBag dataBag, Profile profile) {
```

Comment:
Only thing required here is the `ProfileFields` instance, I'd pass this in instead of the whole profile